tree/treeplayer/inc/TBranchProxy.h

```diff
@@ -160,10 +168,185 @@ namespace Detail {
          }
       }
 
+private:
+      friend class ROOT::Internal::TTreeReaderValueBase;
+
+      enum class EReadType {
+         kDefault,
+         kNoDirector,
+         kReadParentNoCollection,
+         kReadParentCollectionNoPointer,
+         kReadParentCollectionPointer,
+         kReadNoParentNoBranchCountCollectionPointer,
+         kReadNoParentNoBranchCountCollectionNoPointer,
+         kReadNoParentNoBranchCountNoCollection,
+         kReadNoParentBranchCountCollectionPointer,
+         kReadNoParentBranchCountCollectionNoPointer,
+         kReadNoParentBranchCountNoCollection
+      };
+
+      EReadType GetReadType() {
+         if (fParent) {
+            if (!fCollection)
+               return EReadType::kReadParentNoCollection;
+            else {
+               if (IsaPointer())
+                  return EReadType::kReadParentCollectionPointer;
+               else
+                  return EReadType::kReadParentCollectionNoPointer;
+            }
+         } else {
+            if (fBranchCount) {
+               if (fCollection) {
+                  if (IsaPointer()) {
+                     return EReadType::kReadNoParentBranchCountCollectionPointer;
+                  } else {
+                     return EReadType::kReadNoParentBranchCountCollectionNoPointer;
+                  }
+               } else {
+                  return EReadType::kReadNoParentBranchCountNoCollection;
+               }
+
+            } else {
+               if (fCollection) {
+                  if (IsaPointer()) {
+                     return EReadType::kReadNoParentNoBranchCountCollectionPointer;
+                  } else {
+                     return EReadType::kReadNoParentNoBranchCountCollectionNoPointer;
+                  }
+               } else {
+                  return EReadType::kReadNoParentNoBranchCountNoCollection;
+               }
+            }
+         }
+         return EReadType::kDefault;
+      }
+
+      Bool_t ReadNoDirector() {
+         return false;
+      }
```

Comment:
It seems that making `EReadType` work like a bit mask with a bit for each of parent, pointer, branch, collection, etc would making this whole function a single line or a few lines for computing the `OR` of all flags. This would be a lot more efficient than using all of this branching code.